src/test/java/de/tum/in/www1/artemis/util/AbstractArtemisIntegrationTest.java

```diff
@@ -100,6 +101,9 @@ public abstract class AbstractArtemisIntegrationTest implements MockDelegate {
     @SpyBean
     protected ExamAccessService examAccessService;
 
+    @SpyBean
+    protected ExamLiveEventsService examLiveEventsService;
```

Comment:
I don't like this approach. Can you instead verify that the database contains no notifications for the specific exam id?

Reply:
Hmm interesting, that was a no-brainer actually, as we're also checking the instance messages by spying on the service. Is it too much white-box testing or what's your thoughts there? But yea, guess then it would be a full e2e right?

Reply:
We use this approach for other webocket notifications, since their content is not saved and this is the only way of retrieving their content. But this is not the case for these type of messages.

My problem with that is that we don't tests "that no messages got created", but instead that messages got send using this specific method. I can think of multiple cases of wrong server behaviour, that would not be catched by your test.

- If we change this code to only create the database objects and instead use some worker thread to send out notifications in parallel, your test will always work. In fact, if the notification for some reason gets saved to the database but not send via websocket, the test will work but the student will still see this after reloading the page.
- If a future developer directly calls `websocketMessaginService.sendMessage()` your assertion also gets bypassed.

Similarly I also don't like the spy on the ExamAccessService. What if there is a bug inside of the called access check method? Instead there should be tests with users without permission that verify the returned status code.

Reply:
Ok, yea for the first point one still might go through the spied-on method, so that won't be a issue, but yea, one really has to know what the function is doing, and I agree with you. We should either check the database (did that) and the websocket rather than spying on some in-between service. I went fro the DB now.

Reply:
Regarding the `ExamAccessService` spy - I also really don't like that. Thats implementation detail, and should not be tested like that. The integration test should not care about *how* authorization is enforced, it should only test *that* it is enforced, and not care about the how.